api/runnerpool/ch_placer.go

```diff
@@ -0,0 +1,87 @@
+/* The consistent hash ring from the original fnlb.
+   The behaviour of this depends on changes to the runner list leaving it relatively stable.
+*/
+package runnerpool
+
+import (
+	"context"
+	"time"
+
+	"github.com/fnproject/fn/api/common"
+	"github.com/fnproject/fn/api/models"
+
+	"github.com/dchest/siphash"
+	"github.com/sirupsen/logrus"
+)
+
+type chPlacer struct{}
+
+func NewCHPlacer() Placer {
+	logrus.Info("Creating new CH runnerpool placer")
+	return &chPlacer{}
+}
+
+// This borrows the CH placement algorithm from the original FNLB.
+// Because we ask a runner to accept load (queuing on the LB rather than on the nodes), we don't use
+// the LB_WAIT to drive placement decisions: runners only accept work if they have the capacity for it.
+func (p *chPlacer) PlaceCall(rp RunnerPool, ctx context.Context, call RunnerCall) error {
+	// The key is just the path in this case
+	key := call.Model().Path
+	sum64 := siphash.Hash(0, 0x4c617279426f6174, []byte(key))
+	timeout := time.After(call.SlotDeadline().Sub(time.Now()))
+	for {
+
+		select {
+		case <-ctx.Done():
+			return models.ErrCallTimeoutServerBusy
+		case <-timeout:
+			return models.ErrCallTimeoutServerBusy
+		default:
+			runners, err := rp.Runners(call)
+			if err != nil {
+				logrus.WithError(err).Error("Failed to find runners for call")
+			} else {
+				i := int(jumpConsistentHash(sum64, int32(len(runners))))
+				for j := 0; j < len(runners); j++ {
+					r := runners[i]
+
+					placed, err := r.TryExec(ctx, call)
```

Comment:
a note that this is going to flood the first runner for a given key unless TryExec is gating this (didn't look). the naive placer flooded the first node in the list to begin with but anyway, we probably need to prevent this in the starry future to shed load across multiple nodes for a given key with some probability to land on any given node from a subset of runners, hopefully based on something not just completely random.

Reply:
Yeah. At the moment, we're not pulling a Status back to the LB from the runner. By seguing this into the return stream from a call placement, we can keep the Runner model on the LB up-to-date with a picture of load; that's the other piece that prevents this swamping.

(The difference between this and the old fnlb is that we can push Status messages every second or so as a part of that response, meaning that an LB has a more up-to-date picture of the runner state than it did if it had to wait on XXX-Fnlb-wait.)